### PR TITLE
Fix quote & splice typing

### PR DESCRIPTION
**Pull request by jbachurski**
Based on Leo's PRs, combining #4911 and #4912 into one. Adds some fixes and refactors based on past review and my own, and adds (hopefully exhaustive) tests.

The most significant fix is the implementation of `mcomp` for quotes and splices based on our discussions. It's tricky as we want to take into account both flexibility and aliasability, and triggering all the possible `mcomp` cases requires constructing corner-case scenarios.

Each of the new test files tests one of the type inference primitives. 
Note that this moves the existing `moregen` tests into their own file.

The commit structure should be helpful fro reviewing, as it has logically independent pieces. Non-equivalent changes are labeled `Fix`, while (hopefully) equivalent ones are `Refactor`s.

Subsumes #4911 and #4912

### CODE REVIEW

File: testsuite/tests/quotation/typing/quotes_splices.ml [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+ flags = "-extension runtime_metaprogramming";
+ expect;
+*)
+
+#syntax quotations on
+
+let _ : unit = ()
+[%%expect {|
+- : unit = ()
+|}]
```
Comment (lpw25) [inline testsuite/tests/quotation/typing/quotes_splices.ml]: Unclear what this test is for

File: testsuite/tests/quotation/typing/quotes_splices.ml [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+
+(* Cancellations with different numbers of layers *)
+
+(* one *)
```
Comment (lpw25) [inline testsuite/tests/quotation/typing/quotes_splices.ml]: Minor nit, it is less jarring if comments are correctly capitalized.

File: testsuite/tests/quotation/typing/quotes_splices.ml [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+
+(* Flexibility checks -- unifying variable under quotes/splices *)
+
+(* one *)
+
+let _ : <[ $('a) ]> expr = <[()]>
+[%%expect {|
+- : <[unit]> expr = <[()]>
+|}]
```
Comment (lpw25) [inline testsuite/tests/quotation/typing/quotes_splices.ml]: I think this test would be slightly better if it had another occurence of `'a` that didn't get unified with anything, so we could see what had happened to the variable too.

File: testsuite/tests/quotation/typing/quotes_splices.ml [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+[%%expect {|
+- : <[<[<[unit]> expr]> expr]> expr = <[<[<[()]>]>]>
+|}]
+
+
+(* Flexibility checks -- unifying locally-equated type under quotes/splices *)
+
+(* FIXME: these are quote-miskinded, but we don't have quoted kinds yet anyway *)
+
+let f (x : Obj.t) (Equal : (<[$Obj.t * int]>, <[<[string]> * int]>) Type.eq) : <[<[string]>]> = x
```
Comment (lpw25) [inline testsuite/tests/quotation/typing/quotes_splices.ml]: It would probably be easier to follow this test if you actually made an abstract `M.t` that was equal to `<[<[string]>]>` under the hood and used that. That possibly relies on making expect tests treat module definitions as top-level, but maybe we should just do that?

File: testsuite/tests/quotation/typing/quotes_splices.ml [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+[%%expect {|
+Line 1, characters 17-22:
+1 | let f (x : int) (Equal : (<[$int * int]>, <[<[string]> * int]>) Type.eq) : <[<[string]>]> = x
+                     ^^^^^
+Error: This pattern matches values of type
+         "(<[$(int) * int]>, <[$(int) * int]>) Type.eq"
+       but a pattern was expected which matches values of type
+         "(<[$(int) * int]>, <[<[string]> * int]>) Type.eq"
+       Type "$(int)" is not compatible with type "<[string]>"
+|}]
```
Comment (lpw25) [inline testsuite/tests/quotation/typing/quotes_splices.ml]: This is the only test showing an incompatibility between two types. Probably we should have enough of these to cover whatever code we have implemented in `mcomp`. It would also be nice to have a comment explaining which incompatibility this is testing and why it is safe for it to be incompatible.

File: testsuite/tests/quotation/typing/quotes_splices.ml [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+(* FIXME: these are quote-miskinded, but we don't have quoted kinds yet anyway *)
+
+let f (x : Obj.t) (Equal : (<[$Obj.t * int]>, <[<[string]> * int]>) Type.eq) : <[<[string]>]> = x
+[%%expect {|
+val f :
+  Obj.t ->
+  (<[$(Obj.t) * int]>, <[<[string]> * int]>) Type.eq -> <[<[string]>]> =
+  <fun>
+|}]
```
Comment (lpw25) [inline testsuite/tests/quotation/typing/quotes_splices.ml]: It would be better if each of these tests had a comment above them describing which unification was being done and which equation it was supposed to produce.

File: testsuite/tests/quotation/typing/quotes_splices.ml [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+[%%expect {|
+- : unit = ()
+|}]
+
+
+(* Cancellations with different numbers of layers *)
+
+(* one *)
+
+let _ : <[$unit]> = ()
```
Comment (lpw25) [inline testsuite/tests/quotation/typing/quotes_splices.ml]: I think it would be useful to add a comment above each test with the relevant equation being tested, e.g.
```ocaml
(* <[$A]> = A *)
```

File: testsuite/tests/quotation/typing/typing.ml [hunk 8]
```diff
@@ -317,7 +317,17 @@ val foo7 : 'b -> <[$('a) -> $('a)]> expr = <fun>
...
+Error: Values do not match:
+         val foo7' : <[$('_a) -> $('_a)]> expr
+       is not included in
+         val foo7' : <[$('_a) -> $('_a)]> expr
+       The type "<[$('_a) -> $('_a)]> expr" is not compatible with the type
+         "<[$('_a) -> $('_a)]> expr"
+       Type "$('_a) -> $('_a)" is not compatible with type
+         "$('_a) -> $('_a)" = "$('_a) -> $('_a)"
+       Type "'_a" is not compatible with type "'_a" = "'_a"
+       The type variable "'a" occurs inside "'a"
```
Comment (lpw25) [inline testsuite/tests/quotation/typing/typing.ml]: This looks like a bug in the occurs check

File: testsuite/tests/quotation/typing/quotes_splices/cancellation.ml [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+|}]
+(* $(<[A]>) = A *)
+let _ : <[ $(<[unit]>) ]> expr = <[()]>
+[%%expect {|
+- : <[unit]> expr = <[()]>
+|}]
+
+(* Two cancellations *)
+
+(* <[$A]> = A *)
```
Comment (lpw25) [inline testsuite/tests/quotation/typing/quotes_splices/cancellation.ml]: The comment is a little confusing. I'd probably replace it with the full equality being tested i.e. ` <[<[$$A]>]> = A`.

File: testsuite/tests/quotation/typing/quotes_splices/moregen.ml [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+       Values do not match:
+         val f : <[($('a) -> unit) * ($('a) -> unit)]> expr
+       is not included in
+         val f : <[(string -> unit) * (int -> unit)]> expr
+       The type "<[(string -> unit) * (string -> unit)]> expr"
+       is not compatible with the type
+         "<[(string -> unit) * (int -> unit)]> expr"
+       Type "(string -> unit) * (string -> unit)" is not compatible with type
+         "(string -> unit) * (int -> unit)" = "(string -> unit) * (int -> unit)"
+       Type "string" = "string" is not compatible with type "int"
```
Comment (lpw25) [inline testsuite/tests/quotation/typing/quotes_splices/moregen.ml:107]: I think we shouldn't be getting the `"string" = "string"` part, it should just say `"string"`. Probably points to a bug around detecting when expansion has done something interesting or not in the presence of quotation and splice cancellation.

File: testsuite/tests/quotation/typing/quotes_splices/unify.ml [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+  type 'a t = <[int -> int]>
+end
+(* We wrap some constructs in [A.t] so that they are checked
+   in [mcomp], but not [unify]. *)
+(* CR quoted-kinds jbachurski: We might need extra versions of A
+   with 'a quoted-kinded. *)
+module A = struct
+  type _ t  = |
+  type _ t' = |
+end
```
Comment (lpw25) [inline testsuite/tests/quotation/typing/quotes_splices/unify.ml]: This one isn't used in this file

File: testsuite/tests/quotation/typing/quotes_splices/moregen.ml (hunk 10 above)
Comment (jbachurski) [reply testsuite/tests/quotation/typing/quotes_splices/moregen.ml:107]: Maybe you're fine with keeping this as-is, since the expansion code is getting replaced in #5402, which fixes these expansions? See [here](https://github.com/oxcaml/oxcaml/pull/5402/changes#diff-082f384e3d17e1f2a22b61482b9ef57ba3ff6d543770e8b0b433fa1af9f9e829L107).

File: testsuite/tests/quotation/typing/typing.ml (hunk 8 above)
Comment (jbachurski) [reply testsuite/tests/quotation/typing/typing.ml]: It was another one of these top-level-only bugs that shows up in expect tests. It seems that one of the `'a` on either side had a quote-splice pair on it, so it's only a trivial unification `'a ~ 'a` if we cancel first. I'm not sure how to trigger this in a test consistently (it has to happen in a `moregen` check where both sides are identical but they end up differing by quotes/splices). This can only happen in the fast `Tvar` path of `moregen`, and so I added a cancellation call there.

File: testsuite/tests/quotation/typing/quotes_splices/moregen.ml (hunk 10 above)
Comment (jbachurski) [reply testsuite/tests/quotation/typing/quotes_splices/moregen.ml]: I looked into it and I'm not sure why it looks this way, but it _does_ fix itself in the follow-up.

Comment (jbachurski) [reply testsuite/tests/quotation/typing/quotes_splices/moregen.ml:107]: I've looked into it and I think this is because til now quote-splice cancellations happen (redundantly) in `Oprint`, so in reality both sides of that `=` differ by some quotes and splices. This stopped happening in the eval PR, as I started reducing types in `Printtyp`. I'd probably rather keep that fix there. Do you think that's fine?

File: testsuite/tests/quotation/typing/typing.ml (hunk 8 above)
Comment (jbachurski) [reply testsuite/tests/quotation/typing/typing.ml]: Actually -- we had definitely already found this at some point. I had reported the related bug in upstream and I remember you provided the example: https://github.com/ocaml/ocaml/issues/14343
We might actually always want to expand here, then. But only doing cancellation keeps changes limited to just metaprogramming.

Comment (lpw25) [reply testsuite/tests/quotation/typing/typing.ml]: I actually think the problem here is slightly different from the upstream one because here I think we are failing the normal occurs check in `moregen` rather than the `moregen_occur` check. I think a better fix than what you have here would be to do what happens in `unify`: guard the fast path with a `deep_occur` check. That should also be sufficient to actually just remove the call to `occur_for Moregen` in the fast path -- which you will notice is missing from the slow path, which should make the code faster than what is there now too.

File: typing/ctype.ml [hunk 12]
```diff
@@ -5430,6 +5514,10 @@ let rec moregen inst_nongen variance type_pairs env t1 t2 =
   try
     match (get_desc t1, get_desc t2) with
       (Tvar { jkind }, _) when may_instantiate inst_nongen t1 ->
+        let t2 =
+          try try_quote_splice_cancel t2
+          with Cannot_expand -> t2
+        in
```
Comment (lpw25) [inline typing/ctype.ml]: I think a better fix is to guard this fast path with a call to `deep_occur` so that we fall back to the slow path if there are any mentions of `t1` in `t2`. That also allows us to remove the `occur_for Moregen` call below.

File: testsuite/tests/quotation/typing/typing.ml (hunk 8 above)
Comment (jbachurski) [reply testsuite/tests/quotation/typing/typing.ml]: Ack, done as suggested.

File: typing/ctype.ml (hunk 12 above)
Comment (jbachurski) [reply typing/ctype.ml]: Ack, done.

Comment (jbachurski) [reply typing/ctype.ml]: The occurs check in the fast path was introduced here: https://github.com/oxcaml/oxcaml/commit/24dc684f007f92f9548746e87462a527017c2d9b
